RZCardEntry/RZCardEntry/CardType.swift

```diff
@@ -0,0 +1,203 @@
+
+        case .diners:       prefix = ["300"..."305", "309", "38"..."39"]
+                            length = [14]
+
+        case .jcb:          prefix = ["3528"..."3589"]
+                            length = [16]
+        }
+
+        var card = ValidationRequirement()
+        card.prefixes = prefix
+        card.lengths = length
+        return card
+    }
+
+    var segmentGroupings: [Int] {
+        switch self {
+        case .amex:      return [4, 6, 5]
+        case .diners:    return [4, 6, 4]
+        default:         return [4, 4, 4, 4]
+        }
+    }
+
+    var maxLength: Int {
+        return validationRequirements.lengths.max() ?? 16
+    }
+
+    var cvvLength: Int {
+        switch self {
+        case .amex: return 4
+        default: return 3
+        }
+    }
+
+    func valid(_ accountNumber: String) -> Bool {
+        return validationRequirements.valid(accountNumber) && CardType.luhnCheck(accountNumber)
+    }
+
+    func prefixValid(_ accountNumber: String) -> Bool {
+        return validationRequirements.prefixValid(accountNumber)
+    }
+
+}
+
+fileprivate extension CardType {
+
+    struct ValidationRequirement {
+        var prefixes = [PrefixContainable]()
+        var lengths = [Int]()
+
+        func valid(_ accountNumber: String) -> Bool {
+            return lengthValid(accountNumber) && prefixValid(accountNumber)
+        }
+
+        func prefixValid(_ accountNumber: String) -> Bool {
+            guard prefixes.count > 0 else { return true }
+            return prefixes.contains { $0.prefixMatches(accountNumber) }
+        }
+
+        func lengthValid(_ accountNumber: String) -> Bool {
+            guard lengths.count > 0 else { return true }
+            return lengths.contains { accountNumber.characters.count == $0 }
+        }
+    }
+
+    // from: https://gist.github.com/cwagdev/635ce973e8e86da0403a
+    static func luhnCheck(_ cardNumber: String) -> Bool {
+        var sum = 0
+        let reversedCharacters = cardNumber.characters.reversed().map { String($0) }
+        for (idx, element) in reversedCharacters.enumerated() {
+            guard let digit = Int(element) else { return false }
+            switch ((idx % 2 == 1), digit) {
+            case (true, 9): sum += 9
+            case (true, 0...8): sum += (digit * 2) % 9
+            default: sum += digit
+            }
+        }
+        return sum % 10 == 0
+    }
+
+}
+
+//MARK: - CardState
+
+enum CardState {
+    case identified(CardType)
+    case indeterminate([CardType])
+    case invalid
+}
+
+extension CardState: Equatable {}
+func ==(lhs: CardState, rhs: CardState) -> Bool {
+    switch (lhs, rhs) {
+    case (.invalid, .invalid): return true
+    case (.indeterminate, .indeterminate): return true
+    case (let .identified(card1), let .identified(card2)): return card1 == card2
+    default: return false
+    }
+}
+
+extension CardState {
+
+    static func fromNumber(_ cardNumber: String) -> CardState {
```

Comment:
#? `fromNumber` `String` is understandable, but a bit odd. Perhaps worth discussing offline.